cookbooks/continuous_delivery.rst

```diff
@@ -0,0 +1,245 @@
+
+* Fully automated using our `integrations`_
+* `Custom implementation`_ using the
+  `Symfony CLI <https://symfony.com/download>`_ commands.
+
+Integrations
+------------
+
+SymfonyCloud provides integrations with the main Git providers of the market.
+Those integrations rely on webhooks those providers trigger on specific events
+such as code push, branch creation or pull request opening: environments are
+created and deleted on the fly when branches are created or deleted, code and
+configuration changes are deployed automatically on pushes. Depending on the
+configuration on SymfonyCloud side, only specific branches can be deployed or
+pull/merge requests can be deployed or skipped.
+
+.. tip::
+
+   To ease reviewers works, for most Git providers, integrations push links to
+   the deployed environments as a commit status or as a comment to the
+   discussion.
+
+Please refer to :doc:`our integrations list </integrations/index>` to discover
+the complete list of the available integrations and their specific configuration.
+
+Custom implementation
+---------------------
+
+For some projects, it might be more appropriate to use a custom implementation.
+With a custom implementation one project can have a semi-automated or a
+completely automated delivery pipeline based on the requirements. In this
+context, the final logic will depend on the project requirements but you will
+find bellow a list of best practices, tips and pointers to help you in your
+implementation.
+
+* `Requirements <cd-requirements>`_
+* `Create and use a machine user for authentication <machine-user>`_
+* `CLI flags and environment variables <cd-env-flags>`_
+* `Creating and deleting environments <cd-env-create-delete>`_
+* `URLs <cd-url>`_
+
+.. _cd-requirements:
+
+Requirements
+^^^^^^^^^^^^
+
+The system implementing a custom SymfonyCloud deployment automation will have
+several requirements:
+
+#. Have the `Symfony CLI <https://symfony.com/download>`_ installed and runnable
+#. Have Git installed and runnable
+#. Have `authentication`_ completed
+
+Authentication
+^^^^^^^^^^^^^^
+
+If the automation is expected to be run interactively by human beings, they must
+first - locally on their machine - setup Git and run ``symfony login`` as
+well as ``symfony account:ssh:key:add``.
+
+.. _machine-user:
+
+If the implementation is to be run in a non-interactive environment (in a system
+like Jenkins, Travis, GitLab CI or GitHub Actions for example), we recommend the
+use of a "Machine user":  head to `SymfonyConnect <https://connect.symfony.com>`_
+and register a new dedicated account.
+
+As this context will not be interactive, interactive authentication will not be
+possible, this is why you have to use an :doc:`API token
+</cookbooks/api_tokens>`.
+
+Additionally, as deployment are based on Git, you will
+have to generate a new SSH key, and register it in SymfonyCloud:
+
+.. code-block:: terminal
+
+   $ ssh-keygen -N '' -f symfonycloud-ci.key
```

Comment:
I'm unsure why, but CircleCI wouldn't let this private key be added via their UI. The error was,

> It looks like this private key is invalid. Please check if you have pasted the right key. If you are sure your key is valid, please contact support for assistance.

When I used this, it worked:

`ssh-keygen -m PEM -t rsa -C "me@example.com" -f symfonycloud-ci.key`